### PR TITLE
Use `ec_group_do_inverse_ord()` in SM2

### PR DESCRIPTION
**Pull request by romen**
This PR build on top of #6116 : 
- d6db841580 moves the `ec_group_do_inverse_ord()` declaration from `crypto/ec/ec_lcl.h` to a newly created `crypto/include/internal/ec_int.h`. The new header is meant to include *private* but *inter-submodule* accessible functions: up until now the SM2 submodule was using only public functions from the EC submodule.
- 313e72e39a updates the SM2 implementation to use `ec_group_do_inverse_ord()` instead of `BN_mod_inverse()`, mitigating a SCA leakage.

*It's 1.1.1 only.*

#### TODOs:
- [x] reshape commit `AuthorDate`s so github history is coherent with git commit history
- [x] gather feedback on the new commits
- [x] add documentation for `ec_group_do_inverse_ord()` in `crypto/include/internal/ec_int.h`
- [x] decide on documentation text: https://github.com/openssl/openssl/pull/6521#discussion_r197213329
- [x] rebase once #6116 is merged
- [x] rebase on master to fix some [-Wstrict-prototype] warnings
- [x] remove "WIP:" prefix for proper review

### CODE REVIEW

**Comment by romen** [comment]
Ping to @mattcaswell and @bbbrumley

File: crypto/sm2/sm2_sign.c [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 #include "internal/sm2.h"
 #include "internal/sm2err.h"
+#include "../ec/ec_lcl.h" /* ec_group_do_inverse_ord() */
```
Comment (romen) [inline crypto/sm2/sm2_sign.c]: @mattcaswell is this the right way to include stuff from the private API of the EC submodule?

Comment (mattcaswell) [reply crypto/sm2/sm2_sign.c]: No...if there are functions shared between internal sub-modules then they should be put in a header file in crypto/include/internal. You might have to create a new one if it doesn't already exist.

Comment (romen) [reply crypto/sm2/sm2_sign.c]: @mattcaswell So probably this latter change needs to be applied to #6116 before merging?

Comment (romen) [reply crypto/sm2/sm2_sign.c]: I mean, including a definition of `ec_group_do_inverse_ord()` in a header in `crypto/include/internal`

Comment (mattcaswell) [reply crypto/sm2/sm2_sign.c]: Well, you could do it either way. In some ways it makes more sense to do it here, i.e. don't make it into an internal include file until you actually need to use it. But whatever you think is easiest.

Comment (dot-asm) [reply crypto/sm2/sm2_sign.c]: Right, concern here is that relative path in #include can vary between implementations. To counter specifically this concern one can add `INCLUDE[sm2_sign.o]=../ec` to crypto/sm2/build.info and include ec_lcl.h.

Comment (mattcaswell) [reply crypto/sm2/sm2_sign.c]: > Right, concern here is that relative path in #include can vary between implementations.

That is *a* concern, but it is not my main concern. As a general principle we should not expose all of the internals of one sub-module to other sub-modules. There should be a clearly defined "API" between one sub-module and another. The internal include files define what it is reasonable for other sub-modules to have access to, and what it is not.

For that reason I do not think we should take that approach.

Comment (romen) [reply crypto/sm2/sm2_sign.c]: Ok, so let's get #6116 merged, given that it's ready, and then I'll move the prototype declaration where is more appropriate in this PR!

Thanks for the feedback!

**Comment by romen** [comment]
I just rebased on the latest #6116 developments and added a new commit (2a6bdb3) which moves the `ec_group_do_inverse_ord()` declaration from `crypto/ec/ec_lcl.h` to a newly created `crypto/include/internal/ec_int.h`. The new header is meant to include private but inter-submodule accessible functions.

I then rebased the old commit that was including `crypto/ec/ec_lcl.h` breaking submodule separation, to instead include the new inter-submodule header.

As a side-note: *In my mind these 2 commits shouldn't be squashed together before merging.*

@mattcaswell @dot-asm would you care to provide more feedback?
- does the new header look good to you?
- does it need extra (`#ifndef`/`#define`) guards?
- are there conventions I am unaware of for naming these guards? 
- does this fix your previous concern(s)?
- is the build system tracking header files too? should the new internal header be listed somewhere?

**Comment by romen** [comment]
*Added documentation for `ec_group_do_inverse_ord()` in `crypto/include/internal/ec_int.h`.*

File: crypto/include/internal/ec_int.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * [1,EC_GROUP::order).
+ *
+ *         res := x^(-1) (mod EC_GROUP::order)
+ *
+ * - When the EC group order is even, this function returns an error.
+ * - When the EC group order is otherwise composite, the correctness of
+ *   of the output is not guaranteed.
+ * - When x is outside the range [1, EC_GROUP::order), the correctness
+ *   of the output is not guaranteed.
+ */
```
Comment (romen) [inline crypto/include/internal/ec_int.h]: @bbbrumley Should I just go ahead and simply say:
```
 * This function expects that the EC_GROUP order is prime and that x is
 * included in the range [1, EC_GROUP::order).
 * In any other case, the correctness of its output is not guaranteed.
```

Comment (romen) [reply crypto/include/internal/ec_int.h]: This should be the contract between the EC module and other submodules about this function, given that other submodules should probably not even know (or threat as an opaque object) what an `EC_METHOD` is and if it has a specialized implementation with looser requirements or not.

**Comment by romen** [comment]
Rebasing this now that #6116 is merged.

File: crypto/include/internal/ec_int.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (romen) [inline crypto/include/internal/ec_int.h]: @mattcaswell Is the copyright line ok? Or should it be
~~~
 * Copyright 2018-2018 The OpenSSL Project Authors. All Rights Reserved.
~~~

Comment (richsalz) [reply crypto/include/internal/ec_int.h]: Just one year is fine.  Our script updates the copyrights on all files touched anyway so don't worry too much.

File: crypto/include/internal/ec_int.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+#ifndef HEADER_OSSL_EC_INTERNAL_H
+# define HEADER_OSSL_EC_INTERNAL_H
+# include <openssl/opensslconf.h>
+
+# ifndef OPENSSL_NO_EC
+
+#  ifdef __cplusplus
+extern "C" {
+#  endif
```
Comment (dot-asm) [inline crypto/include/internal/ec_int.h]: This is internal header and it's never exposed to C++ compiler, hence there is no need for this. But if one chooses to have it, wouldn't after #include <openssl/ec.h> be the right spot?

Comment (romen) [reply crypto/include/internal/ec_int.h]: > This is internal header and it's never exposed to C++ compiler, hence there is no need for this.

That's a very good point, and one I wondered about myself: I went for inclusion when I copied the basic structure form the recent `crypto/include/internal/sm2.h`.

My final rationale in keeping also what didn't make much sense to me was that this header was very recently added after a long review, and that reviewer definitely knew better!

Comment (mattcaswell) [reply crypto/include/internal/ec_int.h]: > My final rationale in keeping also what didn't make much sense to me was that this header was very recently added after a long review, and that the reviewers definitely knew better!

No, the reviewers just missed it! :-(

Comment (romen) [reply crypto/include/internal/ec_int.h]: @dot-asm I put #6554 to fix also the SM2 internal header

File: crypto/include/internal/ec_int.h (hunk 3 above)
Comment (romen) [reply crypto/include/internal/ec_int.h]: Thanks @richsalz

File: crypto/include/internal/ec_int.h (hunk 4 above)
Comment (romen) [reply crypto/include/internal/ec_int.h]: @mattcaswell It's not like reviewers are reviewing tons of code these days!! 😛 

BTW, I'm trying out the `git commit --fixup` trick you suggested: thanks, that's very useful!!

Comment (dot-asm) [reply crypto/include/internal/ec_int.h]: That specific header was not added, but moved from public. Well, it's not really an excuse, just an explanation...

File: crypto/include/internal/ec_int.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Internal EC functions for other submodules: not for application use */
+
+#ifndef HEADER_OSSL_EC_INTERNAL_H
+# define HEADER_OSSL_EC_INTERNAL_H
```
Comment (mattcaswell) [inline crypto/include/internal/ec_int.h]: Side note: We don't seem to have a consistent naming scheme for these guards. We should look at that - but this looks ok for now.

File: crypto/include/internal/ec_int.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ * Computes the multiplicative inverse of x in the range
+ * [1,EC_GROUP::order).
+ *
+ *         res := x^(-1) (mod EC_GROUP::order)
+ *
+ * - When the EC group order is even, this function returns an error.
+ * - When the EC group order is otherwise composite, the correctness
+ *   of the output is not guaranteed.
+ * - When x is outside the range [1, EC_GROUP::order), the correctness
+ *   of the output is not guaranteed.
```
Comment (mattcaswell) [inline crypto/include/internal/ec_int.h]: We should probably state what the return values are

File: crypto/include/internal/ec_int.h [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ */
+int ec_group_do_inverse_ord(const EC_GROUP *group, BIGNUM *res,
```
Comment (mattcaswell) [inline crypto/include/internal/ec_int.h]: Would it be worth adding `__owur` here? (warn on unused result)

File: crypto/include/internal/ec_int.h (hunk 6 above)
Comment (romen) [reply crypto/include/internal/ec_int.h]: I'll address return values as soon as I get feedback from @bbbrumley on https://github.com/openssl/openssl/pull/6521#discussion_r197213329.

He has a better understanding than me about the non-default paths when `EC_METHOD`s provide their own specific implementation of this functionality.

What I wrote here (or the alternative text) is definitely correct for the new default path, but given that the specialized versions have been written without documentation I don't want to make wrong assumptions: he looked at that code more than I did!

File: crypto/include/internal/ec_int.h (hunk 7 above)
Comment (romen) [reply crypto/include/internal/ec_int.h]: I'd say that annotation is a very good idea, always!

I'll look up how you are adding it on other functions to try and be consistent with the code style!

**Comment by romen** [comment]
About lifting the WIP, I'd like to finalize the documentation on the header before that::we might possibly need to discuss if the "contract" I wrote is too restrictive considering existing implementations and cornercase applications, and decide how to address that.

File: crypto/include/internal/ec_int.h (hunk 2 above)
Comment (bbbrumley) [reply crypto/include/internal/ec_int.h]: I would say `If the EC_GROUP order is even` so there is no ambiguity of which "group" we're referring to. `group->order` is the cardinality of the subgroup generated by G.

**Comment by romen** [comment]
@mattcaswell @dot-asm @richsalz I have lifted the WIP, it should be now ready for proper review.

@bbbrumley could you once again comment on the current text for the documentation?

File: crypto/include/internal/ec_int.h (hunk 6 above)
Comment (romen) [reply crypto/include/internal/ec_int.h]: Latest revision explicitly states the return values and the caveats on reported success.

File: crypto/include/internal/ec_int.h (hunk 7 above)
Comment (romen) [reply crypto/include/internal/ec_int.h]: Added in the last revision.

**Comment by dot-asm** [comment]
Merged. Thanks!

**Pull request closed without merging**